File: content/problems/0106-construct-binary-tree-from-inorder-and-postorder-traversal/solutions/reverse-postorder-recursion/Solution.java
class Solution {
    private int postIndex; private Map<Integer, Integer> positions;
    public TreeNode buildTree(int[] inorder, int[] postorder) { positions = new HashMap<>(); for (int index = 0; index < inorder.length; index++) positions.put(inorder[index], index); postIndex = postorder.length - 1; return build(0, inorder.length - 1, postorder); }
    private TreeNode build(int left, int right, int[] postorder) { if (left > right) return null; int value = postorder[postIndex--], middle = positions.get(value); TreeNode root = new TreeNode(value); root.right = build(middle + 1, right, postorder); root.left = build(left, middle - 1, postorder); return root; }
}

File: content/problems/0106-construct-binary-tree-from-inorder-and-postorder-traversal/solutions/reverse-postorder-recursion/solution.cpp
class Solution {
    int postIndex; unordered_map<int, int> positions;
    TreeNode* build(int left, int right, vector<int>& postorder) { if (left > right) return nullptr; int value = postorder[postIndex--], middle = positions[value]; TreeNode* root = new TreeNode(value); root->right = build(middle + 1, right, postorder); root->left = build(left, middle - 1, postorder); return root; }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) { for (int index = 0; index < (int)inorder.size(); ++index) positions[inorder[index]] = index; postIndex = postorder.size() - 1; return build(0, inorder.size() - 1, postorder); }
};
